File: src/main/scala/com/nekokittygames/Thaumic/Tinkerer/common/core/misc/MobAspects.scala
package com.nekokittygames.Thaumic.Tinkerer.common.core.misc

import net.minecraft.entity.Entity
import net.minecraft.entity.monster._
import net.minecraft.entity.passive._
import net.minecraft.tileentity.TileEntity
import thaumcraft.api.aspects.{Aspect, AspectList}
import thaumcraft.common.entities.monster.{EntityWisp, EntityFireBat, EntityBrainyZombie}

import scala.collection.mutable
import scala.reflect.internal.util.Collections

/**
  * Created by katsw on 24/11/2015.
  */
object MobAspects {

  val aspects=new mutable.HashMap[Class[_ <: Entity],Array[Aspect]]
  // Vanilla Entities
  aspects.put(classOf[EntitySnowman],Array[Aspect](Aspect.WATER,Aspect.WATER,Aspect.MAN))
  aspects.put(classOf[EntityBat],Array[Aspect](Aspect.AIR,Aspect.AIR,Aspect.FLIGHT))
  aspects.put(classOf[EntityBlaze],Array[Aspect](Aspect.FIRE,Aspect.FIRE,Aspect.FIRE))
  aspects.put(classOf[EntityCaveSpider],Array[Aspect](Aspect.BEAST,Aspect.UNDEAD,Aspect.DEATH))
  aspects.put(classOf[EntityCaveSpider],Array[Aspect](Aspect.BEAST,Aspect.UNDEAD,Aspect.DEATH))
  aspects.put(classOf[EntityChicken],Array[Aspect](Aspect.FLIGHT,Aspect.BEAST,Aspect.PLANT))
  aspects.put(classOf[EntityCow],Array[Aspect](Aspect.BEAST,Aspect.EARTH,Aspect.BEAST))
  aspects.put(classOf[EntityCreeper],Array[Aspect](Aspect.ELDRITCH,Aspect.ENERGY,Aspect.BEAST))
  aspects.put(classOf[EntityEnderman],Array[Aspect](Aspect.ELDRITCH,Aspect.ELDRITCH,Aspect.MAN))
  aspects.put(classOf[EntityGhast],Array[Aspect](Aspect.FIRE,Aspect.FLIGHT,Aspect.FLIGHT))
  aspects.put(classOf[EntityHorse],Array[Aspect](Aspect.BEAST,Aspect.BEAST,Aspect.MOTION))
  aspects.put(classOf[EntityIronGolem],Array[Aspect](Aspect.METAL,Aspect.METAL,Aspect.MAN))
  aspects.put(classOf[EntityMagmaCube],Array[Aspect](Aspect.FIRE,Aspect.DARKNESS,Aspect.DARKNESS))
  aspects.put(classOf[EntityMooshroom],Array[Aspect](Aspect.BEAST,Aspect.EARTH,Aspect.LIFE))
  aspects.put(classOf[EntityOcelot],Array[Aspect](Aspect.BEAST,Aspect.EARTH,Aspect.ELDRITCH))
  aspects.put(classOf[EntityPig],Array[Aspect](Aspect.BEAST,Aspect.EARTH,Aspect.MOTION))
  aspects.put(classOf[EntityPigZombie],Array[Aspect](Aspect.UNDEAD,Aspect.FIRE,Aspect.MAN))
  aspects.put(classOf[EntitySheep],Array[Aspect](Aspect.EARTH,Aspect.TOOL,Aspect.BEAST))
  aspects.put(classOf[EntitySilverfish],Array[Aspect](Aspect.METAL,Aspect.METAL,Aspect.EARTH))
  aspects.put(classOf[EntitySkeleton],Array[Aspect](Aspect.UNDEAD,Aspect.MAN,Aspect.UNDEAD))
  aspects.put(classOf[EntitySlime],Array[Aspect](Aspect.DARKNESS,Aspect.DARKNESS,Aspect.BEAST))
  aspects.put(classOf[EntitySpider],Array[Aspect](Aspect.BEAST,Aspect.UNDEAD,Aspect.UNDEAD))
  aspects.put(classOf[EntitySquid],Array[Aspect](Aspect.WATER,Aspect.WATER,Aspect.WATER))
  aspects.put(classOf[EntityVillager],Array[Aspect](Aspect.MAN,Aspect.MAN,Aspect.MAN))
  aspects.put(classOf[EntityWolf],Array[Aspect](Aspect.BEAST,Aspect.BEAST,Aspect.BEAST))
  aspects.put(classOf[EntityZombie],Array[Aspect](Aspect.UNDEAD,Aspect.UNDEAD,Aspect.UNDEAD))

  // TODO: Add more creatures


  // Thaumcraft Entities
  aspects.put(classOf[EntityBrainyZombie],Array[Aspect](Aspect.ELDRITCH,Aspect.UNDEAD,Aspect.MAN))
  aspects.put(classOf[EntityFireBat],Array[Aspect](Aspect.FLIGHT,Aspect.FIRE,Aspect.ELDRITCH))
  aspects.put(classOf[EntityWisp],Array[Aspect](Aspect.AIR,Aspect.ELDRITCH,Aspect.ELDRITCH))


  def getAspects(clazz:Class[_]):Array[Aspect] =
  {
    if(aspects.contains(clazz.asInstanceOf[Class[_ <: Entity]]))
      {
        aspects.get(clazz.asInstanceOf[Class[_ <: Entity]]).get
      }
    else
      {
        null
      }
  }

  def getClazz(inAspect:Array[Aspect]):Class[_]=
  {
    for((key:Class[_],value:Array[Aspect])<-aspects)
      {
        if(value.forall(x=> inAspect.contains(x)))
          {
            return key
          }
      }
    return null
  }
}

File: src/main/java/com/nekokittygames/thaumictinkerer/api/MobAspects.java
/*
 * Copyright (c) 2020. Katrina Knight
 */

package com.nekokittygames.thaumictinkerer.api;

import com.nekokittygames.thaumictinkerer.ThaumicTinkerer;
import net.minecraft.crash.CrashReport;
import net.minecraft.entity.Entity;
import net.minecraft.entity.monster.*;
import net.minecraft.entity.passive.*;
import net.minecraft.util.ReportedException;
import net.minecraft.world.World;
import net.minecraftforge.fml.common.ObfuscationReflectionHelper;
import org.jetbrains.annotations.Contract;
import thaumcraft.api.aspects.Aspect;
import thaumcraft.api.aspects.AspectList;
import thaumcraft.common.entities.monster.EntityBrainyZombie;
import thaumcraft.common.entities.monster.EntityFireBat;
import thaumcraft.common.entities.monster.EntityWisp;

import java.lang.reflect.Method;
import java.util.HashMap;

public class MobAspects {
    private static  final HashMap<Class<?>,MobAspect> aspects=new HashMap<>();

    @Contract(pure = true)
    public static  HashMap<Class<?>, MobAspect> getAspects() {
        return aspects;
    }

    static {
        putAspect(EntitySnowman.class,"SnowMan",new AspectList().add(Aspect.WATER,2).add(Aspect.MAN,1));
        putAspect(EntityBat.class,"Bat",new AspectList().add(Aspect.AIR,2).add(Aspect.FLIGHT,1));
        putAspect(EntityBlaze.class,"Blaze",new AspectList().add(Aspect.FIRE,3));
        putAspect(EntityBrainyZombie.class,"BrainyZombie",new AspectList().add(Aspect.MAGIC,1).add(Aspect.UNDEAD,1).add(Aspect.DARKNESS,1),"thaumcraft");
        putAspect(EntityFireBat.class,"Firebat",new AspectList().add(Aspect.FLIGHT,1).add(Aspect.FIRE,1).add(Aspect.MAGIC,1),"thaumcraft");
        putAspect(EntityCaveSpider.class,"CaveSpider",new AspectList().add(Aspect.BEAST,1).add(Aspect.DEATH,2));
        putAspect(EntityChicken.class,"Chicken",new AspectList().add(Aspect.LIFE,1).add(Aspect.FLIGHT,1).add(Aspect.BEAST,1));
        putAspect(EntityCow.class,"Cow",new AspectList().add(Aspect.BEAST,2).add(Aspect.EARTH,1));
        putAspect(EntityCreeper.class,"Creeper",new AspectList().add(Aspect.MAGIC,1).add(Aspect.BEAST,1).add(Aspect.ELDRITCH,1));
        putAspect(EntityEnderman.class,"Enderman",new AspectList().add(Aspect.ELDRITCH,2).add(Aspect.MAN,1),0.3f,0.0f);
        putAspect(EntityGhast.class,"Ghast",new AspectList().add(Aspect.FIRE,1).add(Aspect.FLIGHT,2),0.1f,0.2f);
        putAspect(EntityHorse.class,"Horse",new AspectList().add(Aspect.BEAST,2).add(Aspect.AIR,1));
        putAspect(EntityIronGolem.class,"VillagerGolem",new AspectList().add(Aspect.METAL,2).add(Aspect.MAN,1),0.3f,0.0f);
        putSlimeAspect(EntityMagmaCube.class,"LavaSlime",new AspectList().add(Aspect.FIRE,1).add(Aspect.WATER,2));
        putAspect(EntityMooshroom.class, "MushroomCow",new AspectList().add(Aspect.BEAST,1).add(Aspect.EARTH,1).add(Aspect.PLANT,1));
        putAspect(EntityOcelot.class,"Ozelot",new AspectList().add(Aspect.BEAST,1).add(Aspect.EARTH,1).add(Aspect.ELDRITCH,1));
        putAspect(EntityPig.class,"Pig",new AspectList().add(Aspect.BEAST,1).add(Aspect.EARTH,1).add(Aspect.MOTION,1));
        putAspect(EntityPigZombie.class, "PigZombie",new AspectList().add(Aspect.UNDEAD,1).add(Aspect.BEAST,1).add(Aspect.FIRE,1));
        putAspect(EntitySheep.class,"Sheep",new AspectList().add(Aspect.EARTH,2).add(Aspect.BEAST,1));
        putAspect(EntitySilverfish.class, "Silverfish",new AspectList().add(Aspect.METAL,2).add(Aspect.EARTH,1));
        putAspect(EntitySkeleton.class, "Skeleton",new AspectList().add(Aspect.UNDEAD,1).add(Aspect.MAN,1).add(Aspect.TOOL,1));
        putSlimeAspect(EntitySlime.class, "Slime",new AspectList().add(Aspect.WATER,2).add(Aspect.BEAST,1));
        putAspect(EntitySpider.class, "Spider",new AspectList().add(Aspect.BEAST,1).add(Aspect.UNDEAD,2));
        putAspect(EntitySquid.class, "Squid",new AspectList().add(Aspect.WATER,3));
        putAspect(EntityWisp.class, "Wisp",new AspectList().add(Aspect.AIR,1).add(Aspect.MAGIC,2),"thaumcraft");
        putAspect(EntityWitch.class, "Witch",new AspectList().add(Aspect.MAGIC,1).add(Aspect.UNDEAD,1).add(Aspect.ELDRITCH,1));
        putAspect(EntityWolf.class, "Wolf",new AspectList().add(Aspect.BEAST,3));
        putAspect(EntityZombie.class, "Zombie",new AspectList().add(Aspect.MAN,1).add(Aspect.UNDEAD,2));
        //put
    }


    public static void checkAspects() {
        for(MobAspect mobAspect: aspects.values()) {
            AspectList curAspects=mobAspect.getAspects();
            for(MobAspect mobAspect2: aspects.values()) {
                if(mobAspect.equals(mobAspect2))
                    continue;
                AspectList checkAspects = mobAspect2.getAspects();
                boolean same=true;
                for(Aspect aspect:curAspects.getAspects()) {
                    if(checkAspects.getAmount(aspect)!=curAspects.getAmount(aspect))
                        same=false;
                }
                if(same)
                    ThaumicTinkerer.logger.error("Warning, Two entities have the same mob aspects"+mobAspect.toString()+", "+mobAspect2.toString());

            }
        }
        ThaumicTinkerer.logger.info("Checked Mob Aspects");
    }

    public static MobAspect getByAspects(AspectList aspectList) {
        for(MobAspect mobAspect2: aspects.values()) {
            AspectList checkAspects = mobAspect2.getAspects();
            boolean same = true;
            for (Aspect aspect : aspectList.getAspects()) {
                if (checkAspects.getAmount(aspect) != aspectList.getAmount(aspect))
                    same = false;
            }
            if (same)
                return mobAspect2;
        }
        return null;
    }
    static void putAspect(Class<?> clazz, String entityName,AspectList aspectsList) {
        aspects.put(clazz,new MobAspect(clazz,entityName,aspectsList));
    }
    static void putAspect(Class<?> clazz, String entityName,AspectList aspectsList,String prefix) {
        aspects.put(clazz,new MobAspect(clazz,entityName,aspectsList).setPrefix(prefix));
    }
    static void putAspect(Class<?> clazz, String entityName,AspectList aspectsList,float scale, float offset) {
        aspects.put(clazz,new MobAspect(clazz,entityName,aspectsList).setScale(scale).setOffset(offset));
    }
    private static final Method setSlimeSizeMethod;

    static
    {
        setSlimeSizeMethod = ObfuscationReflectionHelper.findMethod(EntitySlime.class, "func_70799_a", void.class, int.class, boolean.class);
    }
    static void putSlimeAspect(Class<?> clazz, String entityName,AspectList aspectsList) {
        aspects.put(clazz,new MobAspect(clazz,entityName,aspectsList) {
            @Override
            public Entity createEntity(World worldObj) {
                Entity entity = super.createEntity(worldObj);
                if(entity instanceof EntitySlime) {
                    try
                    {
                        setSlimeSizeMethod.invoke(entity, 1, true);

                    }
                    catch (ReflectiveOperationException e)
                    {
                        throw new ReportedException(new CrashReport("Could not call method '" + setSlimeSizeMethod.getName() + "'", e));
                    }
                }
                return entity;
            }
        });
    }
}
